src/software/ai/hl/stp/tactic/attacker/attacker_fsm_test.cpp

```diff
@@ -0,0 +1,64 @@
+#include "software/ai/hl/stp/tactic/attacker/attacker_fsm.h"
+
+#include <gtest/gtest.h>
+
+#include "software/test_util/test_util.h"
+
+TEST(AttackerFSMTest, test_transitions)
+{
+    World world = ::TestUtil::createBlankTestingWorld();
+    Robot robot = ::TestUtil::createRobotAtPos(Point(-2, -3));
+    Pass pass   = Pass(Point(0, 0), Point(2, 0), 5);
+
+    AttackerFSM::ControlParams control_params{
+        .pass                   = std::make_optional<Pass>(pass),
+        .shot                   = std::nullopt,
+        .chip_target            = std::nullopt,
+        .attacker_tactic_config = std::make_shared<AttackerTacticConfig>()};
+
+    FSM<AttackerFSM> fsm;
+    EXPECT_TRUE(fsm.is(boost::sml::state<DribbleFSM>));
+
+    // robot far from attacker point
+    fsm.process_event(AttackerFSM::Update(
+        control_params, TacticUpdate(robot, world, [](std::unique_ptr<Intent>) {})));
+    EXPECT_TRUE(fsm.is(boost::sml::state<PivotKickFSM>));
+    EXPECT_TRUE(
+        fsm.is<decltype(boost::sml::state<PivotKickFSM>)>(boost::sml::state<DribbleFSM>));
+
+    // robot close to attacker point
+    robot = ::TestUtil::createRobotAtPos(Point(2, 2));
+    fsm.process_event(AttackerFSM::Update(
+        control_params, TacticUpdate(robot, world, [](std::unique_ptr<Intent>) {})));
+    EXPECT_TRUE(fsm.is(boost::sml::state<PivotKickFSM>));
+    EXPECT_TRUE(
+        fsm.is<decltype(boost::sml::state<PivotKickFSM>)>(boost::sml::state<DribbleFSM>));
+
+    // robot at attacker point and facing the right way
+    robot.updateState(RobotState(pass.passerPoint() - Vector(0.1, 0), Vector(),
+                                 pass.passerOrientation(), AngularVelocity::zero()),
+                      Timestamp::fromSeconds(0));
+
+    // process event once to fall through the Dribble FSM
+    fsm.process_event(AttackerFSM::Update(
+        control_params, TacticUpdate(robot, world, [](std::unique_ptr<Intent>) {})));
+    EXPECT_TRUE(fsm.is(boost::sml::state<PivotKickFSM>));
+    EXPECT_TRUE(
+        fsm.is<decltype(boost::sml::state<PivotKickFSM>)>(boost::sml::state<DribbleFSM>));
+
+    // robot should now kick the ball
+    fsm.process_event(AttackerFSM::Update(
+        control_params, TacticUpdate(robot, world, [](std::unique_ptr<Intent>) {})));
+    EXPECT_TRUE(fsm.is(boost::sml::state<PivotKickFSM>));
+    EXPECT_TRUE(fsm.is<decltype(boost::sml::state<PivotKickFSM>)>(
+        boost::sml::state<PivotKickFSM::KickState>));
+
+    // FSM should be done now after 2 ticks
```

Comment:
Can you explain why the FSM needs two more ticks to complete, as in what's being processed in each tick?

Reply:
it's because of the subFSMs, updating comments